src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp

```diff
@@ -0,0 +1,531 @@
+
+            if (total_size / max_block_size < num_streams)
+                num_streams = static_cast<size_t>(total_size / max_block_size);
+
+            if (num_streams == 0)
+                num_streams = 1;
+
+            /// Ranges state, all streams will share the state.
+            auto ranges_state = std::make_shared<NumbersRangedSource::RangesState>();
+            for (size_t i = 0; i < num_streams; ++i)
+            {
+                auto source = std::make_shared<NumbersRangedSource>(intersected_ranges, ranges_state, max_block_size);
+
+                if (i == 0)
+                    source->addTotalRowsApprox(total_size);
+
+                pipe.addSource(std::move(source));
+            }
+            return pipe;
+        }
+    }
+
+    /// Fall back to NumbersSource
+    for (size_t i = 0; i < num_streams; ++i)
+    {
+        auto source
+            = std::make_shared<NumbersSource>(max_block_size, numbers_storage.offset + i * max_block_size, num_streams * max_block_size);
+
+        if (numbers_storage.limit && i == 0)
+        {
+            auto rows_appr = *(numbers_storage.limit);
+            if (limit > 0 && limit < rows_appr)
+                rows_appr = limit;
+            source->addTotalRowsApprox(rows_appr);
+        }
+
+        pipe.addSource(std::move(source));
+    }
+
+    if (numbers_storage.limit)
+    {
+        size_t i = 0;
+        auto storage_limit = *(numbers_storage.limit);
+        /// This formula is how to split 'limit' elements to 'num_streams' chunks almost uniformly.
+        pipe.addSimpleTransform(
+            [&](const Block & header)
+            {
+                ++i;
+                return std::make_shared<LimitTransform>(header, storage_limit * i / num_streams - storage_limit * (i - 1) / num_streams, 0);
+            });
+    }
+
+    return pipe;
+}
+
+ActionsDAGPtr ReadFromSystemNumbersStep::buildFilterDAG()
+{
+    std::unordered_map<std::string, ColumnWithTypeAndName> node_name_to_input_node_column;
+    return ActionsDAG::buildFilterActionsDAG(filter_nodes.nodes, node_name_to_input_node_column, context);
+}
+
+void ReadFromSystemNumbersStep::checkLimits(size_t rows)
+{
+    const auto & settings = context->getSettingsRef();
+
+    SizeLimits limits;
+    if (settings.read_overflow_mode == OverflowMode::THROW && settings.max_rows_to_read)
+        limits = SizeLimits(settings.max_rows_to_read, 0, settings.read_overflow_mode);
+
+    SizeLimits leaf_limits;
+    if (settings.read_overflow_mode_leaf == OverflowMode::THROW && settings.max_rows_to_read_leaf)
+        leaf_limits = SizeLimits(settings.max_rows_to_read_leaf, 0, settings.read_overflow_mode_leaf);
+
+    if (limits.max_rows || leaf_limits.max_rows)
+    {
+        limits.check(rows, 0, "rows (controlled by 'max_rows_to_read' setting)", ErrorCodes::TOO_MANY_ROWS);
+        leaf_limits.check(rows, 0, "rows (controlled by 'max_rows_to_read_leaf' setting)", ErrorCodes::TOO_MANY_ROWS);
+    }
```

Comment:
I think we don't need the extra condition.
```suggestion
    if (settings.read_overflow_mode == OverflowMode::THROW && settings.max_rows_to_read)
    {
        const auto limits = SizeLimits(settings.max_rows_to_read, 0, settings.read_overflow_mode);
        limits.check(rows, 0, "rows (controlled by 'max_rows_to_read' setting)", ErrorCodes::TOO_MANY_ROWS);
    }
    
    if (settings.read_overflow_mode_leaf == OverflowMode::THROW && settings.max_rows_to_read_leaf)
    {
        const auto leaf_limits = SizeLimits(settings.max_rows_to_read_leaf, 0, settings.read_overflow_mode_leaf);
        leaf_limits.check(rows, 0, "rows (controlled by 'max_rows_to_read_leaf' setting)", ErrorCodes::TOO_MANY_ROWS);
    }
```